sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,123 +32,206 @@
 
 import org.joda.time.Instant;
 
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+
 /**
  * Determine the timing and other properties of a new pane for a given computation, key and window.
  * Incorporates any previous pane, whether the pane has been produced because an
  * on-time {@link AfterWatermark} trigger firing, and the relation between the element's timestamp
  * and the current output watermark.
  */
 public class PaneInfoTracker {
-  private TimerInternals timerInternals;
 
-  public PaneInfoTracker(TimerInternals timerInternals) {
-    this.timerInternals = timerInternals;
+  /**
+   * Interface for calculating and tracking pane info state.
+   */
+  public interface PaneInfoTrackerIface {
+    void clear(StateAccessor<?> state);
+
+    ReadableState<PaneInfo> getNextPaneInfo(
```

Comment:
Since you are breaking this out, it would be nicest to take a window, key, and `StateAccessor` separately rather than coupling to the `ReduceFn.Context`.